### DISCUSSION TITLE
Certificates not generating

### DISCUSSION THREAD

**Discussion by Venkatesh-Dayani23** [Sunbird-RC/community#838] [Issues] [answered]
Hi Team,
We are upgrading from 4.6.0 to 5.0.0 Sunbird version. As part of upgrade we have deployed Sunbird-RC services as well. when we try to generate the certificates we are getting below error logs in the **Registry** service of Sunbird-RC.

_registry-5f8b6f6f47-ch8x8 registry 2023-10-04T15:28:26.984805443+05:30 09:58:26.983 [http-nio-8081-exec-3] INFO  d.s.elastic.ElasticServiceImpl:110 - connection info: index:attestationpolicy connectioninfo:10.2.1.5:9200
registry-5f8b6f6f47-ch8x8 registry 2023-10-04T15:28:27.061070637+05:30 09:58:27.060 [http-nio-8081-exec-3] INFO  d.s.elastic.ElasticServiceImpl:114 - resthighclient obj:org.elasticsearch.client.RestHighLevelClient@5c92f1f9
registry-5f8b6f6f47-ch8x8 registry 2023-10-04T15:28:27.264307922+05:30 09:58:27.263 [http-nio-8081-exec-3] ERROR d.s.r.service.ElasticSearchService:87 - Elastic search operation - {}_

**Can you kindly let us  know** 
1.How to create this index - attestationpolicy and what data is should have ? 
2. How to add public and private keys for certificate sign ?

cc: @raghupathiguduri

**Comment by holashchand** [comment, marked as answer]
Creating the index **`attestationpolicy`** in elastic search
```
curl --location --request PUT 'localhost:9200/attestationpolicy?pretty=null' \
--header 'Content-Type: application/json' \
--data '{
    "settings": {}
}'
```
Additionally, you can refer here https://www.elastic.co/guide/en/elasticsearch/reference/7.17/index-modules.html#index-modules-settings to know more about settings

**Comment by holashchand** [comment]
Adding public key and private key for certificate sign
- To generate new public/private key pair, you can follow [here](https://linuxhint.com/generate-ssh-keys-on-linux/) or other resources available on internet
- After generating key pairs, you can replace `default` issuer in this file which is provided as a volume to `certificate-signer`
- Or you can add new entry in issuers in the same file and use that as an issuer in your credential

**Reply by Venkatesh-Dayani23** [reply]
Hi @holashchand 
We created the index attestationpolicy in elastic search and add the public and private keys in certificateapi-config and certificatesign-config. Still we are getting the below error logs in the registry service.

_12:09:55.828 [main] ERROR d.s.r.service.ElasticSearchService:95 - Exception while auditing org.springframework.beans.factory.BeanCr ││ eationException: Error creating bean with name 'scopedTarget.apiMessage': Scope 'request' is not active for the current thread; con ││ sider defining a scoped proxy for this bean if you intend to refer to it from a singleton; nested exception is java.lang.IllegalSta ││ teException: No thread-bound request found: Are you referring to request attributes outside of an actual web request, or processing ││  a request outside of the originally receiving thread? If you are actually operating within a web request and still receive this me ││ ssage, your code is probably running outside of DispatcherServlet: In this case, use RequestContextListener or RequestContextFilter ││  to expose the current request._ 

_12:09:58.002 [main] INFO  d.s.registry.util.EntityParenter:232 - Adding index to shard: shard1 for definition: Training             ││ 12:09:58.002 [main] INFO  d.s.registry.sink.SqlgProvider:88 - Could not create unique index for empty properties                    ││ 12:09:58.077 [main] ERROR d.s.registry.util.EntityParenter:271 - No value present                                                   ││ 12:09:58.078 [main] ERROR d.s.registry.util.EntityParenter:272 - Failed Transaction creating index Training                         ││ 12:09:58.078 [main] INFO  d.s.registry.util.EntityParenter:232 - Adding index to shard: shard1 for definition: Issuer               ││ 12:09:58.078 [main] INFO  d.s.registry.sink.SqlgProvider:88 - Could not create unique index for empty properties                    ││ 12:09:58.186 [main] ERROR d.s.registry.util.EntityParenter:271 - No value present                                                   ││ 12:09:58.186 [main] ERROR d.s.registry.util.EntityParenter:272 - Failed Transaction creating index Issuer                           ││ 12:09:58.186 [main] INFO  d.s.registry.util.EntityParenter:232 - Adding index to shard: shard1 for definition: Schema               ││ 12:09:58.278 [main] ERROR d.s.registry.util.EntityParenter:271 - No value present                                                   ││ 12:09:58.278 [main] ERROR d.s.registry.util.EntityParenter:272 - Failed Transaction creating index Schema                           ││ 12:09:58.279 [main] INFO  d.s.registry.util.EntityParenter:232 - Adding index to shard: shard1 for definition: Signatory            ││ 12:09:58.279 [main] INFO  d.s.registry.sink.SqlgProvider:88 - Could not create unique index for empty properties_    

**Can you kindly suggest on this please**

cc: @raghupathiguduri

**Reply by holashchand** [reply]
These errors should not create an issue in registry,
1. index errors are due to unavailability of tables, it will not should once you invite some entity and restart.
2. Showing apiMessage and index errors on startup are yet be resolved but should not create an issue.

**Reply by Venkatesh-Dayani23** [reply]
ok @holashchand We are getting the below error logs in the Collection Certificate Generator job. Can you kindly suggest on this as well please.

_2023-10-06 05:30:43,420 ERROR org.sunbird.job.certgen.functions.CertificateGeneratorFunction [] - RC Create Error Response: 500 ::  │
│ Response: {"id":"sunbird-rc.registry.create","ver":"1.0","ets":1696570243315,"params":{"resmsgid":"","msgid":"e2741e38-d512-46fd-95 │
│ 18-8c3abcc53a7e","err":"","status":"UNSUCCESSFUL","errmsg":"Unable to create signature: Unrecognized token 'Error': was expecting ( │
│ JSON String, Number, Array, Object or token 'null', 'true' or 'false')\n at [Source: (String)\"{\"error\": \"Not Found\", \"message │
│ \": Error: Request failed with status code 404\"; line: 1, column: 40]"},"responseCode":"OK","result":{}}                           │
│ 2023-10-06 05:30:43,420 ERROR org.sunbird.job.certgen.functions.CertificateGeneratorFunction [] - certificate rc failed for api: se │
│ rvice.rc.create.api | Status is: 500                                                                                                │
│ 2023-10-06 05:30:43,421 ERROR org.sunbird.job.exception.InvalidEventException              [] - Error while processing message for  │
│ Partition: 0 and Offset: 29. Error : Something Went Wrong While Making API Call:  service.rc.create.api | Status is: 500            │
│ org.sunbird.job.certgen.exceptions.ServerException: Something Went Wrong While Making API Call:  service.rc.create.api | Status is: │
│  500_                             

cc: @raghupathiguduri

**Reply by holashchand** [reply]
please make sure, the credential template being used is correct

**Reply by Venkatesh-Dayani23** [reply]
Hi @holashchand 
We are not sure where the credential template is located can you kindly share the steps to check it please.

cc: @raghupathiguduri

**Reply by Venkatesh-Dayani23** [reply]
Hi @holashchand 

We are getting the below mentioned error logs in certificate sign service.Can you kindly let us know how to fix this.

_certificatesign-694b94fd69-5rbz4 certificatesign API POST /sign called
certificatesign-694b94fd69-5rbz4 certificatesign (node:30) Warning: Label '/sign' already exists for console.time()
certificatesign-694b94fd69-5rbz4 certificatesign Fetching credential templates:  https://sunbirdsaasdev.blob.core.windows.net/content/schema/credential_template.json
certificatesign-694b94fd69-5rbz4 certificatesign Input received {
certificatesign-694b94fd69-5rbz4 certificatesign   '@context': [
certificatesign-694b94fd69-5rbz4 certificatesign     '[https://sunbirdsaasdev.blob.core.windows.net/content/schema/v1_context.json'](https://sunbirdsaasdev.blob.core.windows.net/content/schema/v1_context.json%27),
certificatesign-694b94fd69-5rbz4 certificatesign     '[https://sunbirdsaasdev.blob.core.windows.net/content/schema/sunbird_context.json'](https://sunbirdsaasdev.blob.core.windows.net/content/schema/sunbird_context.json%27)
certificatesign-694b94fd69-5rbz4 certificatesign   ],
certificatesign-694b94fd69-5rbz4 certificatesign   type: [ 'VerifiableCredential' ],
certificatesign-694b94fd69-5rbz4 certificatesign   id: 'did:sunbird:{{osid}}',
certificatesign-694b94fd69-5rbz4 certificatesign   issuanceDate: '{{osCreatedAt}}',
certificatesign-694b94fd69-5rbz4 certificatesign   credentialSubject: {
certificatesign-694b94fd69-5rbz4 certificatesign     type: '{{certificateLabel}}',
certificatesign-694b94fd69-5rbz4 certificatesign     recipientName: '{{recipient.name}}',
certificatesign-694b94fd69-5rbz4 certificatesign     trainingName: '{{training.name}}',
certificatesign-694b94fd69-5rbz4 certificatesign     trainingId: '{{training.id}}'
certificatesign-694b94fd69-5rbz4 certificatesign   },
certificatesign-694b94fd69-5rbz4 certificatesign   issuer: {
certificatesign-694b94fd69-5rbz4 certificatesign     id: '[https://raw.githubusercontent.com/project-sunbird/sunbird-devops/release-4.8.0/kubernetes/helm_charts/sunbird-RC/registry/templates/READ.md#Issuer'](https://raw.githubusercontent.com/project-sunbird/sunbird-devops/release-4.8.0/kubernetes/helm_charts/sunbird-RC/registry/templates/READ.md#Issuer%27),
certificatesign-694b94fd69-5rbz4 certificatesign     type: [ 'Issuer' ],
certificatesign-694b94fd69-5rbz4 certificatesign     name: '{{issuer.name}}',
certificatesign-694b94fd69-5rbz4 certificatesign     url: '{{issuer.url}}',
certificatesign-694b94fd69-5rbz4 certificatesign     publicKey: [ '{{issuer.kid}}' ]
certificatesign-694b94fd69-5rbz4 certificatesign   }
certificatesign-694b94fd69-5rbz4 certificatesign } {
certificatesign-694b94fd69-5rbz4 certificatesign   training: {
certificatesign-694b94fd69-5rbz4 certificatesign     osid: '1-997f55c3-b927-4d79-abf9-f36c044de375',
certificatesign-694b94fd69-5rbz4 certificatesign     osUpdatedAt: '2023-10-13T05:16:37.811Z',
certificatesign-694b94fd69-5rbz4 certificatesign     osCreatedAt: '2023-10-13T05:16:37.811Z',
certificatesign-694b94fd69-5rbz4 certificatesign     osUpdatedBy: 'anonymous',
certificatesign-694b94fd69-5rbz4 certificatesign     name: 'Tekdi test certificate course sachin 1012',
certificatesign-694b94fd69-5rbz4 certificatesign     osCreatedBy: 'anonymous',
certificatesign-694b94fd69-5rbz4 certificatesign     id: 'do_11390257187405004813799',
certificatesign-694b94fd69-5rbz4 certificatesign     batchId: '0139025798814187520',
certificatesign-694b94fd69-5rbz4 certificatesign     type: 'Course'
certificatesign-694b94fd69-5rbz4 certificatesign   },
certificatesign-694b94fd69-5rbz4 certificatesign   osUpdatedAt: '2023-10-13T05:16:37.811Z',
certificatesign-694b94fd69-5rbz4 certificatesign   osUpdatedBy: 'anonymous',
certificatesign-694b94fd69-5rbz4 certificatesign   signatory: [
certificatesign-694b94fd69-5rbz4 certificatesign     {
certificatesign-694b94fd69-5rbz4 certificatesign       image: 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAHAAAAAuCAYAAADwZJ3MAAAABGdBTUEAALGPC/xhBQAAACBjR9RRyex071v+B1KgcxKJpS7tw3QjIEWEGvjhbnpNJjHbMk/yypRw6lUpXrzOLNZeRysDPtY0HVv0FJ7vPV1XPoB+7j97LqbJklifm89Ca93FTtqM3WGEyyrGzHWP+xT0Jw1ddqGqIJOfScsJDr5CRnIVUavh/75wTAhpNVgyMuKPTySkui+LQydvpGV1AUkIRTzzwR7Qjbvxt70aWJ1bgYKMFQBDEFFcn88YHjyNIdTyX+TKxgeW0dXizJyuT4mI3DFPWbNpYyqY7PkEq0Vvaza9GkDB01Yn61nByChK5ULeEEcEZD6dugoPrMAuiL/yYRQUUBCkVtXHsOZxB4aUYtH16enulJCzpYOcv/oNwdQ06vQ2Zv3mCiXETafHnkYhNTOrt2PfpZvac2sTKhHweWLeHjj4ffv7bFym66INZN46nqonBUWfe2HUPSYkV+Hq00toVQHnNQkyGaVamFeBoP2JJ9xEECQMjXtR0x5FzKZ7KnnBG2gVcRR0kagpJXVRCpH8FjrbDiEVfvItYTECD0Yas49t499CdCCI5tVeU+KiquH3VJM09gTS3qokMqkQ75kJDdyRB6jY04aX0D7rx6gc/o6QuhCe2vk3ELc38cf9DZJ9fSoBvM0//+D3iQsrwdBti18ePcjg7gLEJR97d/wB5JfFEhLRi0I9zq+GiRfw2CwoGtK7UtoVwpmoJpa1JGEUOBLi2cPuiQpLuPsctLtXY2U4iEn19emIRAXVTjryZ9RRHcpfww81ZhHlXsv3RTDbdmc1jO3LZe/wBXnv/YXy92pgWZIxOWHPX2mI6BtW8susRRCIFLz35OqNaJY+9+jsk1np+ev9OVibnorQbAsQcz9/K0YJ0tm7M5sAna+kecuPJ7buJCryEwSjDSj45a/5OTyvoGvSlojmaosbFNPYFMSXIWODdwT3pHxHnV4W/2xXk0nG+bWVo1gUUBAn7j23h+PnFPP/YS6TFnMNkEvPj7YfZty8I7chi1m/IQ60O59DJLSRFFyK30mEQufHs7/+b8MA6fnb3CzS3h/Crt58hJT6PJza/hbeqEYCrY84cPrmKl3Y9jvcCHQXZanIvqrn/wfO4OfYiERuxVhiu04tv4CdSeocWUNYUS37DrdQOBCMxTBN7SxmZq94jyKsGd8dO5JLx67rPrB8jOvpCeer559my7kO2rD74fykyGIxydu+N552/JuLp54R/rDcll0NYFFrOnoMpKAOteHBdFg+v/QNSGWT+8jVkCj3/88TLmA1TVNSFc6kinPLKAHr67XF1GUehdOZCnhdpSRVYO1mhmxTx0jMv4OPWOGP+jU/aU98VyYmy2yisXYRYgLiIWpZGnGahey1uDu2Ib2ABfZYFlPHnD3dw4nQAf37pBZyVw/92VRDEdPZ4Ut0UTVFDKh8eSGSwS2BySoVHsIKkiFJsZMMYsOf0mQic7HpQqwcwGcQ0DfnS32lHRupZMn/wIUF+TZwpTOfVnWt4+7XfolROsf2ZF8m8ezd3ph+5oV6NTaqo7Y3kQpWGuvYghkyeuCmHSV2YT1LgGTxUHYhFM9N3nNUtdNosoazUm8jAqn+KZxakjIyq6OrzoKJRQ019EF0DAdQ3+NHdYMJokpOYWMqjDx5AbjWNeVrKiFZJT4sUiULPfRv2Ud+WgL7Cl42rsvnJpjdxduwFwDBxFVv5CCrlMErHCby8xhkbtbohvuj09rT0BlJQv5SC6gRGzFZ4ug6TrikhPjAfL1UbMsnEjD/TWRVQIhawkk0wOS6jrVtNbauasyWp1LaGMGUS4aYaZHFYJaGB/TR2+2GrdmeiU4Sn+wh3LD+JXPZZ7GpsDeTI6VsJDWpnUKvmbF06P9ywlw0p+1BIx/55v6DgUUavOnLitIY5CYII=',
certificatesign-694b94fd69-5rbz4 certificatesign       osUpdatedAt: '2023-10-13T05:16:37.811Z',
certificatesign-694b94fd69-5rbz4 certificatesign       osCreatedAt: '2023-10-13T05:16:37.811Z',
certificatesign-694b94fd69-5rbz4 certificatesign       osUpdatedBy: 'anonymous',
certificatesign-694b94fd69-5rbz4 certificatesign       name: 'sachin',
certificatesign-694b94fd69-5rbz4 certificatesign       osCreatedBy: 'anonymous',
certificatesign-694b94fd69-5rbz4 certificatesign       designation: 'CEO',
certificatesign-694b94fd69-5rbz4 certificatesign       osid: '1-bf893e11-d031-4f59-a29b-2e82464337c7',
certificatesign-694b94fd69-5rbz4 certificatesign       id: 'CEO/CEO'
certificatesign-694b94fd69-5rbz4 certificatesign     }
certificatesign-694b94fd69-5rbz4 certificatesign   ],
certificatesign-694b94fd69-5rbz4 certificatesign   recipient: {
certificatesign-694b94fd69-5rbz4 certificatesign     osid: '1-599809a5-c1ca-485d-a447-410a03ec9a6e',
certificatesign-694b94fd69-5rbz4 certificatesign     osUpdatedAt: '2023-10-13T05:16:37.811Z',
certificatesign-694b94fd69-5rbz4 certificatesign     osCreatedAt: '2023-10-13T05:16:37.811Z',
certificatesign-694b94fd69-5rbz4 certificatesign     osUpdatedBy: 'anonymous',
certificatesign-694b94fd69-5rbz4 certificatesign     osCreatedBy: 'anonymous',
certificatesign-694b94fd69-5rbz4 certificatesign     name: 'aseeya test',
certificatesign-694b94fd69-5rbz4 certificatesign     id: '35af3208-f312-44b8-9edf-9328c90d0c2a'
certificatesign-694b94fd69-5rbz4 certificatesign   },
certificatesign-694b94fd69-5rbz4 certificatesign   osid: '1-5ec8f99b-4faf-4f6c-ba15-95c763a0b67c',
certificatesign-694b94fd69-5rbz4 certificatesign   osOwner: [ 'anonymous' ],
certificatesign-694b94fd69-5rbz4 certificatesign   templateUrl: '[https://sunbirdsaasdev.blob.core.windows.net/content/content/do_113898436380065792133/artifact/do_113898436380065792133_1696586472448_certificate_2023-10-06_15_31.svg'](https://sunbirdsaasdev.blob.core.windows.net/content/content/do_113898436380065792133/artifact/do_113898436380065792133_1696586472448_certificate_2023-10-06_15_31.svg%27),
certificatesign-694b94fd69-5rbz4 certificatesign   certificateLabel: 'Certificate of merit',
certificatesign-694b94fd69-5rbz4 certificatesign   osCreatedAt: '2023-10-13T05:16:37.811Z',
certificatesign-694b94fd69-5rbz4 certificatesign   osCreatedBy: 'anonymous',
certificatesign-694b94fd69-5rbz4 certificatesign   issuer: {
certificatesign-694b94fd69-5rbz4 certificatesign     osid: '1-f95e74be-48fb-41b6-85d3-8af7e2660bfa',
certificatesign-694b94fd69-5rbz4 certificatesign     osUpdatedAt: '2023-10-13T05:16:37.811Z',
certificatesign-694b94fd69-5rbz4 certificatesign     osCreatedAt: '2023-10-13T05:16:37.811Z',
certificatesign-694b94fd69-5rbz4 certificatesign     osUpdatedBy: 'anonymous',
certificatesign-694b94fd69-5rbz4 certificatesign     kid: 'c08794a5-ced9-40ee-94a6-bb64eea40e6f',
certificatesign-694b94fd69-5rbz4 certificatesign     name: 'NULP',
certificatesign-694b94fd69-5rbz4 certificatesign     osCreatedBy: 'anonymous',
certificatesign-694b94fd69-5rbz4 certificatesign     url: 'https://org_site.org'
certificatesign-694b94fd69-5rbz4 certificatesign   },
certificatesign-694b94fd69-5rbz4 certificatesign   status: 'ACTIVE'
certificatesign-694b94fd69-5rbz4 certificatesign }
certificatesign-694b94fd69-5rbz4 certificatesign Error: You must pass a string or Handlebars AST to Handlebars.compile. You passed [object Object]
certificatesign-694b94fd69-5rbz4 certificatesign     at Object.compile (/app/node_modules/handlebars/dist/cjs/handlebars/compiler/compiler.js:494:11)
certificatesign-694b94fd69-5rbz4 certificatesign     at HandlebarsEnvironment.hb.compile [as _compile] (/app/node_modules/handlebars/dist/cjs/handlebars.js:39:40)
certificatesign-694b94fd69-5rbz4 certificatesign     at HandlebarsEnvironment.Handlebars.compile (/app/node_modules/handlebars-delimiters/index.js:53:32)
certificatesign-694b94fd69-5rbz4 certificatesign     at generateCredentials (/app/src/services/signerService.js:16:33)
certificatesign-694b94fd69-5rbz4 certificatesign     at generateCredentialsRoute (/app/src/routes/signerController.js:14:18)
certificatesign-694b94fd69-5rbz4 certificatesign     at processTicksAndRejections (node:internal/process/task_queues:96:5)
certificatesign-694b94fd69-5rbz4 certificatesign     at async Server.<anonymous> (/app/main.js:41:32) {
certificatesign-694b94fd69-5rbz4 certificatesign   description: undefined,
certificatesign-694b94fd69-5rbz4 certificatesign   fileName: undefined,
certificatesign-694b94fd69-5rbz4 certificatesign   lineNumber: undefined,
certificatesign-694b94fd69-5rbz4 certificatesign   endLineNumber: undefined,
certificatesign-694b94fd69-5rbz4 certificatesign   number: undefined
certificatesign-694b94fd69-5rbz4 certificatesign }_

**Reply by holashchand** [reply]
ok, which version of certificate-signer you are using?
Also please make sure your certificate template url is accessible in the pods
https://sunbirdsaasdev.blob.core.windows.net/content/content/do_113898436380065792133/artifact/do_113898436380065792133_1696586472448_certificate_2023-10-06_15_31.svg%27

**Reply by Venkatesh-Dayani23** [reply]
we are using 5.0.0 certificate-signer and the template file is publicly accessible  

https://sunbirdsaasdev.blob.core.windows.net/content/content/do_113898436380065792133/artifact/do_113898436380065792133_1696586472448_certificate_2023-10-06_15_31.svg

**Reply by holashchand** [reply]
Okay, I think there are some issues with your template, let's connect and see

**Reply by Venkatesh-Dayani23** [reply]
Hi @holashchand 

I have scheduled a call at 3 pm and sent you the invite.Kindly join the meeting at 3 PM.
meet link: https://meet.google.com/rma-vkju-kww

**Reply by Venkatesh-Dayani23** [reply]
Hi @holashchand 
Can you confirm are you joining the call ?

**Reply by Venkatesh-Dayani23** [reply]
Hi @sachinb80,
As discussed with @holashchand we need to edit the Certificate template. Kindly edit the Certificate template.

cc: @raghupathiguduri

**Reply by sachinb80** [reply]
@holashchand Please provide us with the certificate layout if possible

**Reply by sachinb80** [reply]
@holashchand Can ypu please provide us the certificate layout for 5.0 version of sunbird?

**Reply by holashchand** [reply]
Certificate layout is defined by the implementer only, I have given all the details required to Venkatesh

**Reply by vinod303** [reply]
@coolbung  @mohit  @rhwarrier 
As per the @holashchand's suggestion, we have now using a different cert template layout as per the credential_template.json.
Still the certificate is not getting generated.

We had a long session debugging session  with @srprasanna  along with Devops team.  Checked RC logs, config files etc.. and also re-deployed RC. Still the certificate is not getting generated. 
 
So his suggestion was to have a combined debugging session with RC team, SBLern and SBSAAS  devops team to figure this out.

@mohit @rhwarrier  seeking the team's  help here. This is happening in SB Ver5.0  which we wanted to upgrade in SB SaaS. This is blocking us from doing so! 
@Venkatesh-Dayani23 @holashchand

**Reply by vinod303** [reply]
@Venkatesh-Dayani23 @raghupathiguduri  can you  pls the post the reason for this and how it is fixed and post that pls mark this  resolved/closed.
